### PR TITLE
Progress on Fairplay support, #382.

### PR DESCRIPTION
**Pull request by chrisfillmore**

### CODE REVIEW

File: lib/hls/hls_parser.js
Comment (chrisfillmore) [inline lib/hls/hls_parser.js]: I believe this is only the case in `com.apple.fps.1_0`.

File: lib/hls/hls_parser.js
Comment (chrisfillmore) [inline lib/hls/hls_parser.js]: I don't know if there is a method for providing init data via an HLS manifest.

File: lib/media/drm_engine.js
Comment (chrisfillmore) [inline lib/media/drm_engine.js]: The demo in the FPS SDK does this, but I believe it is application-specific business logic. None of the Fairplay license proxies I work with expect this behaviour from clients.

File: lib/polyfill/patchedmediakeys_safari.js
Comment (chrisfillmore) [inline lib/polyfill/patchedmediakeys_safari.js]: This isn't necessary for `com.apple.fps.2_0`

File: lib/polyfill/patchedmediakeys_safari.js
Comment (chrisfillmore) [inline lib/polyfill/patchedmediakeys_safari.js]: Here as well, I've never seen CKC-wrapped license responses from the servers I've worked with. The application can handle this via response filters.

File: lib/polyfill/patchedmediakeys_safari.js
Comment (chrisfillmore) [inline lib/polyfill/patchedmediakeys_safari.js]: The application should still have a way of getting the `contentId`, but I think it could be an event from the player.

**Comment by chrisfillmore** [comment]
I've been able to complete a license request, but haven't actually gotten playback yet. My next steps in testing will be to do a custom build and integrate it with our custom app to see the results.

File: lib/polyfill/patchedmediakeys_safari.js
Comment (TheModMaker) [inline lib/polyfill/patchedmediakeys_safari.js]: In a real app, where does the certificate come from?  Can the license (or another) server handle this message and give us the certificate?  This happens in the Chrome CDM where the first request may be a certificate request and the license server is expected to give it to us.  It would be nice to avoid a special case for both us and the app.

Comment (chrisfillmore) [reply lib/polyfill/patchedmediakeys_safari.js]: I am not a Fairplay / license proxy expert so I asked our CWIP guy down the hall, and he tells me that it is not comparable to the Widevine flow. The Fairplay license proxy expects that the license challenge is already signed with the certificate.

Comment (chrisfillmore) [reply lib/polyfill/patchedmediakeys_safari.js]: Is your concern that the app will now be required to provide a certificate to Shaka for Fairplay playback, whereas it is optional for Widevine? Is this what you mean by "special case"?

EME says:

>**setServerCertificate**
Provides a server certificate to be used to encrypt messages to the license server.
>
>Key Systems that use such certificates must also support requesting the certificate from the server via the Queue a "message" Event algorithm.

Since Safari doesn't implement the latest spec, and doesn't have this API, and we can't predict whether Apple will eventually implement it, or if the implementation will follow the spec, it may not be possible to avoid a special case for Safari.

**Comment by tchakabam** [comment]
@chrisfillmore @TheModMaker @joeyparrish Are you guys familiar with this implementation of Fairplay certificate/license flow with the webkit prefixed API? 

---> https://github.com/chemoish/videojs-fairplay

It doesn't use MediaSource for playback, but relies on Safari built-in HLS support (which makes it necessary to side-load webvtt for subtitles).

Anyhow, maybe a good comparison point for seeing how it can be done!

I am also very interested at making this effort here advance, and am happy to contribute any kind of work if needed :hand:

**Comment by chrisfillmore** [comment]
@tchakabam thanks for the share. In general I am familiar with playback via the native Safari player. This uses the keySystem `com.apple.fps.1_0`, which has different behaviour than `com.apple.fps.2_0` (which is required for playback via MSE).

Compared to the work required to play HLS-FP via MSE, playback via the native player is relatively easy. Apple provides a reference implementation in the FairPlay SDK. For this reason I'm of the opinion that Shaka should not support `com.apple.fps.1_0`, but I don't think the wider community agrees.

**Comment by sandersaares** [comment]
As a member of the wider community, I am happy to help paint a picture of desires by stating my disinterest in `fps.1_0` and great interest in `fps.2_0` in all its ways and forms.

Are there contributions besides writing code that might benefit activity on this topic?

**Comment by chrisfillmore** [comment]
@sandersaares Thank you for sharing that, because I'm concerned that client teams misunderstand what they are asking for when they ask Shaka team to support playback via `src=`. The native Safari player is at the same level of abstraction as Shaka player; Shaka should not wrap it. Wrapping it will lead to client teams thinking that playback issues via `src=` are issues with Shaka (instead of issues with Safari).

I am only working on this topic sporadically because my application provides playback in Safari via `src=`, so this isn't high priority for us. If you have some time or resources to contribute to this, what this branch needs is some testing with FP-protected content. The test content I was working with previously was problematic (possibly due to multiplexed content being transmuxed, see #1143). I haven't put the time into testing with our real production content.

**Comment by joeyparrish** [comment]
@chrisfillmore, I agree with you that supporting src= HLS will cause confusion among developers.  I predict bugs asking why the buffering or networking configs have no effect, at the very least.  Other things like seeking behavior will also differ.

@sandersaares, I would also greatly prefer `fps.2_0`, but there is insufficient guidance from Apple on how to use it.  It would also be unusable on iOS due to the lack of MediaSource, so it would only apply to desktop Safari.

There has still been significant interest in a built-in fallback to src=, both for convenience of iOS/Fairplay support (#997 and #382) and for the convenience of plain MP4 playback (#816).  I am nervous about it, but we will try to put appropriate messaging in the JS console to make it clear that this is a convenience feature only, and that much of the player configuration will be ineffective in this mode.

None of this will preclude `fps.2_0` support, if we can figure it out.  It would be preferred to `fps.1_0` whenever possible.

**Comment by tchakabam** [comment]
Totally agree with all you! :) 

Shaka should be a MediaSource data-provider engine in the first place. Most of it's purpose will not be useful on Safari when the built-in HLS streaming engine will be used.

What is needed to do FP 1.0 on Safari via src URL is only a Fairplay license/cert request building - and that is where I thought it might be useful for Shaka to mimic that from the videojs example, but I didn't realize that via MediaSource we need another version of FP.

So what is needed to play FPS HLS on Safari is really mosty something that is built quite easily independently of all the other objects Shaka lib provides imho. But then again, there is a good overlap on dealing with EME and its prefixed APIs etc as we can see :)

Maybe this is also a way to expand the purpose of Shaka to become more general purpose HTML5 playback "toolbox" library? 

For example, one can still record playback stats and currently played resolution on Safari (via video.videoWidth/Height props) even though it's not possible to drive the quality switches ourselves.

I think it will be useful to show certain console errors or throw exceptions when developers try to use Shaka methods that are ineffective on Safari.

@joeyparrish What is the current state here on FPS 1.0 and what is lacking to make it work? :)

**Comment by joeyparrish** [comment]
This PR and the experimental branch it is based on are both driving toward `fps.2_0`.

For `fps.1_0`, the DRM stuff is clearly documented by Apple.  Before we can implement the FairPlay part, we need #997 (src= HLS), and that depends on #816 (src= mp4).  After #816, we will no longer assume MediaSource, and after #997, we will be using Apple's native HLS when necessary.  Then adding `fps.1_0` is relatively easy.

If `fps.2_0` can be solved, there are no other issues blocking its use.

**Comment by mstattma** [comment]
@joeyparrish Do you have any hints on how fps.2_0 licenses look like? How are you testing this license server wise?

**Comment by joeyparrish** [comment]
No, I don't have any information on `fps.2_0` licenses yet.

That castlabs link does not seem to say anything about either MSE or EME, just FairPlay DRM.  My understanding of the news is that iOS now has EME, but still does not have MSE.

**Comment by mstattma** [comment]
Alright, curious to learn if any shareable information surfaces. 

You're right about mobile Safari, sorry for my confusion. I initially thought Apple aligned MSE/EME with regular Safari, but it seems they've just added their proprietary EME for `fps.1_0` to non-MSE playback.

**Comment by chrisfillmore** [comment]
@joeyparrish I think we may be closer to getting this working than you think. The problems I'm blocked by are not related to Fairplay, they're all content-related. I'm not really very familiar with HLS or actual media issues, though. I can contribute time and test content, but is there someone from Shaka team who could work with me to troubleshoot these content issues? I have a contact at Apple on the WebKit team who could unblock us if we run into a wall.

Our content plays fine in Safari via `src=`, but when I try to play in Shaka, the HLS parser throws HLS_COULD_NOT_PARSE_SEGMENT_START_TIME because `hasPesPacket` was false. I haven't filed a ticket because it's cumbersome for us to share our content with industry partners. (The guy sitting next to me has taken over work on #1198 to try to remedy this.)

**Review by crypt096 — APPROVED** [review]

**Comment by joeyparrish** [comment]
@crypt096, please refrain from approving pull requests.  Since you are not a member of this project, I think it sends confusing signals to the authors of the PRs.

**Comment by avelad** [comment]
@joeyparrish I think that can be closed with https://github.com/google/shaka-player/commit/6d76a135e5128dfd47653acea025d0a264d121d5

**Comment by joeyparrish** [comment]
Thanks, @avelad.  Closing.

**Pull request closed without merging**
